enterprise/trackedusers/users_reporter.go

```diff
@@ -0,0 +1,227 @@
+package trackedusers
+
+import (
+	"context"
+	"database/sql"
+	"encoding/hex"
+	"fmt"
+	"time"
+
+	"github.com/rudderlabs/rudder-server/jobsdb"
+	migrator "github.com/rudderlabs/rudder-server/services/sql-migrator"
+	txn "github.com/rudderlabs/rudder-server/utils/tx"
+
+	"github.com/rudderlabs/rudder-go-kit/config"
+	"github.com/rudderlabs/rudder-go-kit/logger"
+	obskit "github.com/rudderlabs/rudder-observability-kit/go/labels"
+
+	"github.com/lib/pq"
+	"github.com/samber/lo"
+	"github.com/segmentio/go-hll"
+	"github.com/spaolacci/murmur3"
+	"github.com/tidwall/gjson"
+)
+
+const (
+	idTypeUserID                = "userID"
+	idTypeAnonymousID           = "anonymousID"
+	idTypeIdentifiedAnonymousID = "identifiedAnonymousID"
+
+	// changing this will be non backwards compatible
+	murmurSeed = 123
+
+	trackUsersTable = "tracked_users_reports"
+)
+
+type UsersReport struct {
+	WorkspaceID              string
+	SourceID                 string
+	UserIDHll                *hll.Hll
+	AnonymousIDHll           *hll.Hll
+	IdentifiedAnonymousIDHll *hll.Hll
+}
+
+//go:generate mockgen -destination=./mocks/mock_user_reporter.go -package=mockuserreporter github.com/rudderlabs/rudder-server/enterprise/trackedusers UsersReporter
+
+// UsersReporter is interface to report unique users from reports
+type UsersReporter interface {
+	ReportUsers(ctx context.Context, reports []*UsersReport, tx *txn.Tx) error
+	GenerateReportsFromJobs(jobs []*jobsdb.JobT, sourceIdFilter map[string]bool) []*UsersReport
+	MigrateDatabase(dbConn string, conf *config.Config) error
+}
```

Comment:
Should we define this interface on the consumer instead?

Do we need the generated mock?

```suggestion
// UsersReporter is interface to report unique users from reports
type UsersReporter interface {
	ReportUsers(ctx context.Context, reports []*UsersReport, tx *txn.Tx) error
	GenerateReportsFromJobs(jobs []*jobsdb.JobT, sourceIdFilter map[string]bool) []*UsersReport
}
```
Should MigrateDatabase be part of the interface for processor?

Reply:
- MigrateDatabase need to be part of this interface as we are running migration as a part of app startup.
- Do we need the generated mock? - removed mock generator
- Should we define this interface on the consumer instead? we need 1 interface here as well as factory can return noop implementation as well, more in the next PR https://github.com/rudderlabs/rudder-server/pull/4827